### ISSUE TITLE
[FEATURE]: Refactor God Classes

### ISSUE THREAD

**Issue by Tgenz1213** [feder-cr/Jobs_Applier_AI_Agent_AIHawk#842]
### Feature summary

Initial discussion for refactoring several design choices to promote better maintainability of the repo

### Feature description

AIHawkEasyApplier and AIHawkJobManager are two god classes that are tightly coupled and have too many responsibilities. While they keep related code in the same place, it is difficult to navigate and make any significant changes. 

### Why does it matter?
1. We can do this voluntarily now, or be forced to do it in the future when the project grows and becomes impossible to manage. 
2. Improved testability and maintainability.
3. It will make development significantly easier by working with smaller files and classes.

### What does "done" look like?
1. All classes have a single responsibility
2. No more tight coupling
3. Basic amount of testing implemented for all new classes
5. All files are organized
6. Updated docs

### Tasks
1. Decide what design features to change
2. Ensure testing for current implementation is complete to prevent regression
3. Ensure there is a plan for rollbacks
4. Update docs, maybe add a simple style guide to help with format/linting config
5. Ensure CI/CD workflows are updated for thorough testing
7. Create tests for new classes 
8. Full-stop on new features and fixing non-urgent bugs, freeze PR submissions, make necessary announcements (may be beneficial to freeze PRs earlier to avoid causing excessive merge conflicts)
9. Incrementally move logic to new classes, starting with logic that is not tightly coupled in the old classes
10. Perform testing
11. Open PRs back up for submission

### Design changes
- POM Pattern for managing pages to separate logic for each page, increasing testability and maintainability. Say LinkedIn takes a page away or adds a new one - we'll have to change one file to adjust while the rest of the pages unphased.
- Repository Pattern to abstract data access from business logic, improving code organization and separation of concerns
- Service Layer Pattern to tie together POM, repository, and LLM integration, eliminating the need for an overly complex Facade Pattern
- Drop Facade Pattern to remove tight coupling between LLM model and other classes
- Separate files for classes that benefit from separation (not entirely necessary, but looks better IMO than having multiple classes in the same file if one doesn't help the other)

### Folder structure
Auto_Jobs_Applier_AI_Agent/
- README.md
- ... (other base-level files, assets/, tests/, .github/, etc.)
- data/
   - example/
      - config.yaml
      - plain_text_resume.yaml
   - config.yaml
   - secrets.yaml
   - plain_text_resume.yaml
   - ...
- docs/
   - style_guide.md
   - architecture_design.md
   - workflow_diagrams.md
   - ...
- output/
   - success.json
   - failed.json
   - skipped.json
   - open_ai_calls.json
- pylint_plugins/
   - __init__.py
   - stylistic_rules.py
   - plugin_config.py (add this to register more than one plugin in .pylintrc)
- src/
   - __init__.py
   - main.py
   - models/
      - llm_model.py (abstract base class)
      - ollama.py
      - perplexity.py
      - openai.py
      - ...
   - pages/
      - __init__.py
      - home_page.py
      - login_page.py
      - job_search_page.py
      - job_posting_page.py
      - easy_apply_upload_page.py
      - easy_apply_question_page.py
      - easy_apply_review_page.py
   - services/
      - job_application_service.py
      - selenium_driver.py
   - repositories/
      - job_repository.py
   - utils/
      - log.py

### Motivation

For better maintainability, easier feature implementations, and debugging - extends the life of the project and invites more collaboration

### Alternatives considered

Other approaches or design patterns may be better than what I recommended. Some changes can be deferred until absolutely necessary.

**Comment by surapuramakhil** [comment]
usefull relevant information 

https://github.com/AIHawk-co/Auto_Jobs_Applier_AI_Agent/pull/833/files/8714460592a0991f86a9ef4aa895f1e482f70e03..78972e61182e217b81b8cea0dfa90bac911a5187?diff=split&w=0

**Comment by Tgenz1213** [comment]
> @Tgenz1213 
> 
> Hi, do you want to work on this?
> 
> For better communication with other open source developers I suggest you join our discord group: https://discord.gg/mMZcMTH9K6

Yes, working on it now. I can share progress when asked. I started a todo list partway through, which won't capture any commits made before I made it. 

It's not perfectly matching this plan so far because I'm tweaking things as I do research to maximize scalability, maintainability, and speed of development.

**Comment by Tgenz1213** [comment]
Please let me know if you have other requests and questions

**Comment by surapuramakhil** [comment]
@Tgenz1213 after looking at proposal, I am afraid more experience required, when it comes to structuring, splitting of files, classes, modularity. It's better to do it in PR reviews.
Another problem is with scope.

can you create a sub takes - let's take only two God classes / files. (Each as separate GitHub issue) you can add more later
(1) job_manager.py
(2) linked_easy_applyer.py

Doing these things in PR reviews is a much better approach than discussing on GitHub issue.
Can you create PR for subtasks? That will be great.

**Comment by Tgenz1213** [comment]
@surapuramakhil the issue is that it requires shotgun surgery to completely abstract these and there's no design that supports how this should be done. It's nearly impossible to solve this with needing to solve other design problems. While my degree maybe isn't enough "experience" I think I can at least present auto applier v5 design docs in a way that makes it modular, maintainable, and scalable. The god classes are one subset of a larger issue that should be all address. 

I'm also drafting a todo list. I'll create a PR with updated docs, target docs, migration docs, and a comprehensive TODO list. I'm almost done with the documentation.

**Comment by surapuramakhil** [comment]
> The` god classes are one subset of a larger issue that should be all addressed

@Tgenz1213 this GitHub issue is all about that. of other things, better to create other GitHub issue for those

You are always encouraged to contribute to any part of this project.
